### PR TITLE
Beautify text file output

### LINKED ISSUE #4: Beautify data generation file output
When performing data generation, the ground truth (and soon to be sensor) data that is generated is outputted to a text file for analysis. However, the current formatting of this written data is not very consistent/stable for larger floats in arrays.

Ex:
image

In that image, the spacing is consistent within each array, but inconsistent as a whole. If an individual number of an array has more characters than the others, the array inserts whitespace to ensure each number within the array has consistent for the entire array. Additionally, if an individual number of an array has more than 10 characters (i.e. `[1000000.0000       0.0000       0.0000]` or `[1000000.0000 1000000.0000 1000000.0000]`), then the spacing between each column of headers/data will be misaligned (due to how the file writes are currently implemented).

The goal here is to be able to write data to file with robust formatting that can handle numbers with large character counts, with consistent spacing, and is easy to read.

### PR DESCRIPTION
**Pull request by dwkletke**
This PR closes #4 and cleans up the data output to text files using the Python `tabulate` module. Currently, the table output is set to `rst` which is similar to reStructuredText format. However, if anyone feels particularly strong about other table styles I am completely open to changing this (list of available styles is [here](https://www.kite.com/python/docs/tabulate.tabulate)).

Additionally as a driveby, I added a check for if the `/state_estimation/generated_files` directory exists and the directory is created if it does not. This is the directory that the generated files go into, and if you didn't already have the directory on your system the code would fail when trying to create the data output files (definitely not intended).

For reference, this is what the output would look like before this PR:
image

This is what the output looks like now:
image

One thing I did not add/change is the precision of the floats in each vector. I am curious if anyone thinks I should set a precision for all of the floats or leave the precision as is.

---
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11)

### CODE REVIEW

File: data_generator.py [hunk 1]
```diff
@@ -114,73 +116,36 @@ def time_update(rocket, time_dict):
...
     time_dict = {"current_time": 0, "end_time": 100, "timestep": 0.01}
 
+    # Data lists and headings initializations
+    gt_gen_data = []
+    sensor_gen_data = []
+    headings_gt = ["Position [m]", "Velocity [m/s]", "Acceleration [m/s^2]", "Orientation"]
+    headings_sd = ["Baro_Pressure [KPa]", "Temperature [Celsius]", "Acceleration [m/s^2]", "Magnetic_Field [T]"]
+
     with open(GT_PATH, "w") as ground_truth:
         with open(SD_PATH, "w") as sensor_data:
```
Comment (usephysics) [inline data_generator.py:133]: nitpicking, these two lines can be combined as
 `with open(GT_PATH, "w") as ground_truth, open(SD_PATH, "w") as sensor_data:`

File: data_generator.py [hunk 2]
```diff
@@ -114,73 +116,36 @@ def time_update(rocket, time_dict):
...
-    sd_file.write(sensor_data_to_write + "\n")
-
-
-# TODO: Check if the title output can be shortened
 def main():
     """
     Main function for generating data based on the input Rocket.
     """
     # Timestep setup
     time_dict = {"current_time": 0, "end_time": 100, "timestep": 0.01}
```
Comment (usephysics) [inline data_generator.py:124]: Is this dictionary necessary? It seems like end_time and timestep are constants. Could all 3 of these values be stored in primitive variables?

**Review by usephysics — COMMENTED** [review]
Should the generated .txt files be added to .gitignore?

**Review by dwkletke — COMMENTED** [review]
I thought about it but I was thinking it might be desirable to keep generated data files in git in the future so we don't need to generate them every time we want to use the program.

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-:-MPVE5_TEUoekqp60O6d:bpqpken)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jacobdeery and @usephysics)

___
*[data_generator.py, line 124 at r1](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-MPCzWyF0pVSveMaAZjr-r1-124:-MPVCN24-gxhJyObvqlE:b-iqy7hm) ([raw file](https://github.com/waterloo-rocketry/state_estimation/blob/d5878a7c5f2e9efe4cabb036d71a1b906c86215d/data_generator.py#L124)):*
<details><summary><i>Previously, usephysics (Ryan Tan) wrote…</i></summary><blockquote>

Is this dictionary necessary? It seems like end_time and timestep are constants. Could all 3 of these values be stored in primitive variables?
</blockquote></details>

You are correct that it isn't necessary. I had grouped the variables together a long time ago since they were related and needed to be passed into the `time_update()` function (with the exception of `end_time`). If you feel like I should remove them from the dictionary then I can do so.
___
*[data_generator.py, line 133 at r1](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-MPCzWyF0pVSveMaAZjr-r1-133:-MPVE2Dj7V9UP069_f-N:b2nhpmf) ([raw file](https://github.com/waterloo-rocketry/state_estimation/blob/d5878a7c5f2e9efe4cabb036d71a1b906c86215d/data_generator.py#L133)):*
<details><summary><i>Previously, usephysics (Ryan Tan) wrote…</i></summary><blockquote>

nitpicking, these two lines can be combined as
 `with open(GT_PATH, "w") as ground_truth, open(SD_PATH, "w") as sensor_data:`
</blockquote></details>

Ty!

**Comment by usephysics** [comment]
> If you feel like I should remove them from the dictionary then I can do so.

I'm unsure if it's bad practice or if it doesn't matter, @jacobdeery probably knows more.

> I thought about it but I was thinking it might be desirable to keep generated data files in git in the future so we don't need to generate them every time we want to use the program.

Perhaps we can put empty text files on the repository and then add to the .gitignore, so anyone cloning the repository has the files already.

Or, leave as is, but then every commit includes some output from whatever test was done last, but this could also serve as a preview. Idk.

**Review by jacobdeery — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jacobdeery and @usephysics)

___
*[data_generator.py, line 124 at r1](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-MPCzWyF0pVSveMaAZjr-r1-124:-MPglvNp7pnnMbSkxNVg:b5vjy29) ([raw file](https://github.com/waterloo-rocketry/state_estimation/blob/d5878a7c5f2e9efe4cabb036d71a1b906c86215d/data_generator.py#L124)):*
<details><summary><i>Previously, dwkletke (Dawson Kletke) wrote…</i></summary><blockquote>

You are correct that it isn't necessary. I had grouped the variables together a long time ago since they were related and needed to be passed into the `time_update()` function (with the exception of `end_time`). If you feel like I should remove them from the dictionary then I can do so.
</blockquote></details>

I would remove them from the dict. I think the only reason why we might want to keep it around is if we think we'll load these from some configuration file in the future, but if that's the case we can bring it back fairly easily.

**Review by dwkletke — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @usephysics)

___
*[data_generator.py, line 124 at r1](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-MPCzWyF0pVSveMaAZjr-r1-124:-MPkbRZh0Wiw4EEnvOIl:b-2rr0tz) ([raw file](https://github.com/waterloo-rocketry/state_estimation/blob/d5878a7c5f2e9efe4cabb036d71a1b906c86215d/data_generator.py#L124)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

I would remove them from the dict. I think the only reason why we might want to keep it around is if we think we'll load these from some configuration file in the future, but if that's the case we can bring it back fairly easily.
</blockquote></details>

Yep sounds good!
___
*[data_generator.py, line 133 at r1](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-MPCzWyF0pVSveMaAZjr-r1-133:-MPkbUuvBul3HeiBmvI6:b-896fix) ([raw file](https://github.com/waterloo-rocketry/state_estimation/blob/d5878a7c5f2e9efe4cabb036d71a1b906c86215d/data_generator.py#L133)):*
<details><summary><i>Previously, dwkletke (Dawson Kletke) wrote…</i></summary><blockquote>

Ty!
</blockquote></details>

Done.

**Review by usephysics — COMMENTED** [review]
should something be done to the sig digs or spacing, so that the rows always have lined up columns? it looks like when scientific notation is used, the row is misaligned

**Review by dwkletke — COMMENTED** [review]
Do you mean the elements of each vector within a column aren't lined up? I mentioned in the initial PR comment that I didn't set any precision for these outputs. However elements in the arrays (for each respective vector) are printed with the same spacing relative to every other element in that array.

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-:-MPlOcEr8xtzErWu_Cww:b3f9irt)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jacobdeery and @usephysics)

**Comment by usephysics** [comment]
Yes, elements of each vector within a column

**Review by dwkletke — COMMENTED** [review]
I set the float precision to an arbitrary number and removed scientific notation but there is still going to be misalignment since the numbers vary in order throughout the simulated flight.

The general idea of making these output files easier to read and manage seems to have been met. This is always something we could look into deeper and fix in another PR.

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/state_estimation/11#-:-MS3XHoL6oMI8kE7YAT3:b-vzxc70)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jacobdeery and @usephysics)

**Review by usephysics — APPROVED** [review]
True, output looks good.

**Pull request merged**